src/lib/rng/chacha_rng/chacha_rng.cpp

```diff
@@ -60,11 +73,18 @@ void ChaCha_RNG::generate_output(std::span<uint8_t> output, std::span<const uint
    }
 
    m_chacha->write_keystream(output);
+
+   // optionally overwrite key after each output operation for backtracking resistance
+   if(m_fast_key_erasure) {
+      const auto next_key = m_chacha->keystream_bytes(m_chacha_keylen);
+      m_chacha->set_key(next_key);
+   }
 }
 
 void ChaCha_RNG::update(std::span<const uint8_t> input) {
    m_hmac->update(input);
-   m_chacha->set_key(m_hmac->final());
+   const auto chacha_key = m_hmac->final();
+   m_chacha->set_key(chacha_key.data(), m_chacha_keylen);
```

Comment:
As mentioned above: an assertion would be nice here. Maybe extract into a helper function and reuse it in both locations.

```C++
namespace {

std::span<const uint8_t> truncate(std::span<const uint8_t> in, size_t bytes) {
   BOTAN_ASSERT_NOMSG(in.size() >= bytes);
   return in.first(bytes);
}

}
```

```C++
void ChaCha_RNG::update(std::span<const uint8_t> input) {
   m_hmac->update(input);
   m_chacha->set_key(truncate(m_hmac->final(), m_chacha_keylen));
   // [...]
}
```